testing/integration_testing/xcode_tool.rb

```diff
@@ -0,0 +1,188 @@
+# associated testapp readme files. This includes adding frameworks, enabling
+# capabilities, and patching files with required data.
+# See the C++ testapp readmes for details on these changes.
+#
+# Usage: xcode_tool.rb [options]
+# -d [dir],               XCode project directory (required)
+#   --XCodeCPP.xcodeProjectDir
+# -t [target],            Assume target to match the .xcodeproj name. (required)
+#   --XCodeCPP.target
+# -f [frameworks],        Paths to the custom frameworks. (required)
+#   --XCodeCPP.frameworks
+# -e [entitlement_path],  Path to entitlements (optional)
+#   --XCodeCPP.entitlement
+
+require 'optparse'
+require 'xcodeproj'
+
+# Performs all the product-specific changes to the xcode project provided by
+# the xcodeProjectDir flag.
+def main
+  OptionParser.new do |opts|
+    opts.banner = 'Usage: xcode_tool.rb [options]'
+    opts.on('-d', '--XCodeCPP.xcodeProjectDir [dir]',
+      'XCode project directory (required)') do |xcode_project_dir|
+      @xcode_project_dir = xcode_project_dir
+    end
+    opts.on('-t', '--XCodeCPP.target [target]',
+      'Assume target to match the .xcodeproj name. (required)') do |target|
+      @target_name = target
+    end
+    opts.on('-f', '--XCodeCPP.frameworks [frameworks]',
+      'Paths to the custom frameworks. (required)') do |frameworks|
+      @frameworks = frameworks.split(",")
+    end
+    opts.on('-e', '--XCodeCPP.entitlement [entitlement_path]',
+      'Path to entitlements (optional)') do |entitlement_path|
+      @entitlement_path = entitlement_path
+    end
+  end.parse!
+
+  raise OptionParser::MissingArgument,'-d' if @xcode_project_dir.nil?
+  raise OptionParser::MissingArgument,'-t' if @target_name.nil?
+  raise OptionParser::MissingArgument,'-f' if @frameworks.nil?
+
+  project_path = "#@xcode_project_dir/#@target_name.xcodeproj"
+  @project = Xcodeproj::Project.open(project_path)
+  @target = @project.targets.first
+
+  # Examine components rather than substrings to minimize false positives.
+  path_components = project_path.split('/')
+  if path_components.include?('FirebaseAuth')
+    make_changes_for_auth
+  elsif path_components.include?('FirebaseMessaging')
+    make_changes_for_messaging
+  elsif path_components.include?('FirebaseDynamicLinks')
+    make_changes_for_dynamiclinks
+  end
+
+  framework_dir = "#@xcode_project_dir/Frameworks"
+  set_build_setting('FRAMEWORK_SEARCH_PATHS', ['${inherited}', framework_dir])
+
+  @frameworks.each do |framework|
+    add_custom_framework(framework)
+  end
+
+  # Bitcode is unnecessary, as we are not submitting these to the Apple store.
+  # Disabling bitcode significantly speeds up builds.
+  set_build_setting('ENABLE_BITCODE', 'NO')
+
+  @project.save
+end
+
+def make_changes_for_auth
+  puts 'Auth testapp detected.'
+  add_entitlements
+  set_reverse_id
+  add_system_framework('UserNotifications')
+  puts 'Finished making auth-specific changes.'
+end
```

Comment:
I'd love to minimize the actual project-specific logic in this script and have it just be a very dumb way for the Python script to set up the project. WDYT?

Reply:
Yeah I strongly agree with that actually, to treat this more like a wrapper for the relevant bits of the xcodeproj library we need and reduce the cases where it needs to be modified. Also to get rid of the error-prone logic to infer the API from the project path, which is just asking for trouble.

Reply:
Agreed - will you make that change here or in a future PR when we get iOS working?

Reply:
I put in a few comments in this PR, but will leave the rest to a future PR - probably shouldn't hold up this week's iteration for this.